pkg/clients/container/cluster.go

```diff
@@ -0,0 +1,996 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package container
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/google/go-cmp/cmp"
+	gke "google.golang.org/api/container/v1beta1"
+
+	"github.com/crossplaneio/stack-gcp/apis/compute/v1beta1"
+	gcp "github.com/crossplaneio/stack-gcp/pkg/clients"
+	container "google.golang.org/api/container/v1beta1"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+)
+
+const (
+	// BootstrapNodePoolName is the name of the node pool that is used to
+	// boostrap GKE cluster creation.
+	BootstrapNodePoolName = "crossplane-bootstrap"
+)
+
+// GenerateNodePoolForCreate inserts the default node pool into
+// *container.Cluster so that it can be provisioned successfully.
+func GenerateNodePoolForCreate(in *container.Cluster) {
+	in.NodePools = []*container.NodePool{
+		&container.NodePool{
+			Name:             BootstrapNodePoolName,
+			InitialNodeCount: 0,
+		},
+	}
+}
+
+// GenerateCluster generates *container.Cluster instance from GKEClusterParameters.
+func GenerateCluster(in v1beta1.GKEClusterParameters) *container.Cluster { // nolint:gocyclo
+	cluster := &container.Cluster{
+		ClusterIpv4Cidr:       gcp.StringValue(in.ClusterIpv4Cidr),
+		Description:           gcp.StringValue(in.Description),
+		EnableKubernetesAlpha: gcp.BoolValue(in.EnableKubernetesAlpha),
+		EnableTpu:             gcp.BoolValue(in.EnableTpu),
+		InitialClusterVersion: gcp.StringValue(in.InitialClusterVersion),
+		LabelFingerprint:      gcp.StringValue(in.LabelFingerprint),
+		Locations:             in.Locations,
```

Comment:
Note: we do not set `Location` here because it is only to be used as a parameter in `Create()`, not in the actual `*container.Cluster` itself (output only).